File: cpp/src/catalog.cpp
#include "forensicator/catalog.hpp"

#include "forensicator/util.hpp"

#include <sqlite3.h>

#include <cstring>
#include <fstream>
#include <sstream>

namespace forensicator {

Statement::Statement(sqlite3* db, const std::string& sql) : db_(db) {
    int rc = sqlite3_prepare_v2(db, sql.c_str(), -1, &stmt_, nullptr);
    if (rc != SQLITE_OK) {
        throw CatalogError(std::string("prepare failed: ") +
                           sqlite3_errmsg(db) + " for: " + sql);
    }
}

Statement::~Statement() {
    if (stmt_) sqlite3_finalize(stmt_);
}

Statement::Statement(Statement&& o) noexcept
    : db_(o.db_), stmt_(o.stmt_) { o.db_ = nullptr; o.stmt_ = nullptr; }

Statement& Statement::operator=(Statement&& o) noexcept {
    if (this != &o) {
        if (stmt_) sqlite3_finalize(stmt_);
        db_ = o.db_;
        stmt_ = o.stmt_;
        o.db_ = nullptr;
        o.stmt_ = nullptr;
    }
    return *this;
}

void Statement::bind_null(int idx) {
    sqlite3_bind_null(stmt_, idx);
}

void Statement::bind(int idx, std::int64_t v) {
    sqlite3_bind_int64(stmt_, idx, v);
}

void Statement::bind(int idx, const std::string& v) {
    sqlite3_bind_text(stmt_, idx, v.data(),
                      static_cast<int>(v.size()), SQLITE_TRANSIENT);
}

void Statement::bind(int idx, std::string_view v) {
    sqlite3_bind_text(stmt_, idx, v.data(),
                      static_cast<int>(v.size()), SQLITE_TRANSIENT);
}

void Statement::bind(int idx, const char* v) {
    sqlite3_bind_text(stmt_, idx, v, -1, SQLITE_TRANSIENT);
}

void Statement::bind_optional(int idx, const std::optional<std::string>& v) {
    if (v.has_value()) bind(idx, *v);
    else bind_null(idx);
}

bool Statement::step() {
    int rc = sqlite3_step(stmt_);
    if (rc == SQLITE_ROW) return true;
    if (rc == SQLITE_DONE) return false;
    throw CatalogError(std::string("step failed: ") + sqlite3_errmsg(db_));
}

void Statement::reset() {
    sqlite3_reset(stmt_);
}

void Statement::clear_bindings() {
    sqlite3_clear_bindings(stmt_);
}

int Statement::col_int(int idx) const {
    return sqlite3_column_int(stmt_, idx);
}

std::int64_t Statement::col_int64(int idx) const {
    return sqlite3_column_int64(stmt_, idx);
}

std::string Statement::col_text(int idx) const {
    const unsigned char* s = sqlite3_column_text(stmt_, idx);
    if (!s) return {};
    int n = sqlite3_column_bytes(stmt_, idx);
    return std::string(reinterpret_cast<const char*>(s), n);
}

std::optional<std::string> Statement::col_text_optional(int idx) const {
    if (sqlite3_column_type(stmt_, idx) == SQLITE_NULL) return std::nullopt;
    return col_text(idx);
}

bool Statement::col_is_null(int idx) const {
    return sqlite3_column_type(stmt_, idx) == SQLITE_NULL;
}

// ---- Catalog ----

Catalog::Catalog() = default;

Catalog::~Catalog() { close(); }

void Catalog::close() {
    if (db_) {
        sqlite3_close(db_);
        db_ = nullptr;
    }
}

void Catalog::probe_locking(const std::filesystem::path& dbpath) {
    // A trivial query to surface broken-locking failures up front.
    char* errmsg = nullptr;
    int rc = sqlite3_exec(db_, "SELECT 1 FROM sqlite_master LIMIT 1;",
                          nullptr, nullptr, &errmsg);
    if (rc != SQLITE_OK) {
        std::string msg = errmsg ? errmsg : "unknown";
        if (errmsg) sqlite3_free(errmsg);
        if (rc == SQLITE_IOERR || rc == SQLITE_BUSY || rc == SQLITE_LOCKED ||
            rc == SQLITE_PROTOCOL || rc == SQLITE_IOERR_LOCK) {
            std::ostringstream oss;
            oss << "cannot open catalog " << dbpath.string() << ": " << msg << "\n"
                << "This filesystem appears to have broken POSIX locking (SMB / NFS / exFAT /\n"
                << "some FUSE mounts). forensicator refuses to operate on such filesystems\n"
                << "because long-running scans corrupt SQLite there (\"database disk image is\n"
                << "malformed\"), and the previous nolock fallback caused real data loss.\n\n"
                << "Move the catalog to a local SSD/HDD with a normal filesystem (APFS / ext4\n"
                << "/ NTFS / HFS+) and re-run. The catalog records mount_point separately so\n"
                << "it can still find files on the original data volume, regardless of where\n"
                << "the SQLite file lives.\n";
            throw CatalogError(oss.str());
        }
        throw CatalogError("cannot open catalog " + dbpath.string() + ": " + msg);
    }
}

void Catalog::open(const std::filesystem::path& path,
                   const std::filesystem::path& schema_sql_path,
                   bool readonly,
                   std::optional<std::string> dataset) {
    int flags = readonly ? SQLITE_OPEN_READONLY
                         : (SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE);
    int rc = sqlite3_open_v2(path.string().c_str(), &db_, flags, nullptr);
    if (rc != SQLITE_OK) {
        std::string err = db_ ? sqlite3_errmsg(db_) : "open failed";
        if (db_) { sqlite3_close(db_); db_ = nullptr; }
        throw CatalogError("cannot open catalog " + path.string() + ": " + err);
    }

    probe_locking(path);

    // Pragmas.
    if (!readonly) {
        char* err = nullptr;
        sqlite3_exec(db_, "PRAGMA journal_mode = WAL;", nullptr, nullptr, &err);
        if (err) { sqlite3_free(err); err = nullptr; }
        sqlite3_exec(db_, "PRAGMA synchronous = NORMAL;", nullptr, nullptr, &err);
        if (err) { sqlite3_free(err); err = nullptr; }
    }
    {
        char* err = nullptr;
        sqlite3_exec(db_, "PRAGMA foreign_keys = ON;", nullptr, nullptr, &err);
        if (err) sqlite3_free(err);
    }
    {
        char* err = nullptr;
        sqlite3_exec(db_, "PRAGMA cache_size = -200000;", nullptr, nullptr, &err);
        if (err) sqlite3_free(err);
    }

    // Check schema_version if meta exists.
    {
        sqlite3_stmt* st = nullptr;
        sqlite3_prepare_v2(db_,
            "SELECT name FROM sqlite_master WHERE type='table' AND name='meta'",
            -1, &st, nullptr);
        bool has_meta = false;
        if (st && sqlite3_step(st) == SQLITE_ROW) has_meta = true;
        if (st) sqlite3_finalize(st);
        if (has_meta) {
            auto sv = get_meta("schema_version");
            int v = sv ? std::stoi(*sv) : 0;
            if (v != SCHEMA_VERSION) {
                throw CatalogError("catalog " + path.string() +
                                   " has schema version " + std::to_string(v) +
                                   ", expected " + std::to_string(SCHEMA_VERSION));
            }
        }
    }

    if (!readonly) {
        init_schema(schema_sql_path, dataset);
    }
}

void Catalog::init_schema(const std::filesystem::path& schema_sql_path,
                          std::optional<std::string> dataset) {
    std::string sql = read_schema_sql(schema_sql_path);
    for (const auto& stmt : split_sql(sql)) {
        if (stmt.find_first_not_of(" \t\r\n") == std::string::npos) continue;
        exec(stmt);
    }
    set_meta("schema_version", std::to_string(SCHEMA_VERSION));
    set_meta("created_at", std::to_string(now_epoch()));
    if (dataset.has_value()) set_meta("dataset_name", *dataset);
}

void Catalog::exec(const std::string& sql) {
    char* err = nullptr;
    int rc = sqlite3_exec(db_, sql.c_str(), nullptr, nullptr, &err);
    if (rc != SQLITE_OK) {
        std::string msg = err ? err : "exec failed";
        if (err) sqlite3_free(err);
        throw CatalogError("sql failed: " + msg + " for: " + sql);
    }
}

void Catalog::begin() { exec("BEGIN"); }
void Catalog::commit() { exec("COMMIT"); }
void Catalog::rollback() { exec("ROLLBACK"); }

std::optional<std::string> Catalog::get_meta(const std::string& key) {
    Statement st(db_, "SELECT value FROM meta WHERE key = ?");
    st.bind(1, key);
    if (!st.step()) return std::nullopt;
    return st.col_text_optional(0);
}

void Catalog::set_meta(const std::string& key, const std::string& value) {
    Statement st(db_, "INSERT OR REPLACE INTO meta(key, value) VALUES(?, ?)");
    st.bind(1, key);
    st.bind(2, value);
    st.step();
}

Volume Catalog::upsert_volume(const std::string& hostname,
                              const std::string& volume,
                              const std::string& mount_point,
                              const std::string& os_n) {
    Statement st(db_,
        "INSERT INTO volumes(hostname, volume, mount_point, os, scanned_at) "
        "VALUES(?,?,?,?,?) "
        "ON CONFLICT(hostname, volume) DO UPDATE SET "
        "mount_point = excluded.mount_point, "
        "os = excluded.os, scanned_at = excluded.scanned_at");
    st.bind(1, hostname);
    st.bind(2, volume);
    st.bind(3, mount_point);
    st.bind(4, os_n);
    st.bind(5, now_epoch());
    st.step();
    auto v = get_volume(hostname, volume);
    if (!v) throw CatalogError("upsert_volume: volume not found after insert");
    return *v;
}

std::optional<Volume> Catalog::get_volume(const std::string& hostname,
                                          const std::string& volume) {
    Statement st(db_,
        "SELECT volume_id, hostname, volume, mount_point, os, scanned_at "
        "FROM volumes WHERE hostname = ? AND volume = ?");
    st.bind(1, hostname);
    st.bind(2, volume);
    if (!st.step()) return std::nullopt;
    Volume v;
    v.volume_id = st.col_int64(0);
    v.hostname = st.col_text(1);
    v.volume = st.col_text(2);
    v.mount_point = st.col_text(3);
    v.os = st.col_text(4);
    v.scanned_at = st.col_int64(5);
    return v;
}

std::vector<Volume> Catalog::volumes() {
    std::vector<Volume> out;
    Statement st(db_,
        "SELECT volume_id, hostname, volume, mount_point, os, scanned_at "
        "FROM volumes ORDER BY hostname, volume");
    while (st.step()) {
        Volume v;
        v.volume_id = st.col_int64(0);
        v.hostname = st.col_text(1);
        v.volume = st.col_text(2);
        v.mount_point = st.col_text(3);
        v.os = st.col_text(4);
        v.scanned_at = st.col_int64(5);
        out.push_back(std::move(v));
    }
    return out;
}

// ---- Schema helpers ----

std::string read_schema_sql(const std::filesystem::path& schema_sql_path) {
    std::ifstream ifs(schema_sql_path, std::ios::binary);
    if (!ifs) {
        throw CatalogError("schema file " + schema_sql_path.string() +
                           ": cannot open");
    }
    std::ostringstream oss;
    oss << ifs.rdbuf();
    return oss.str();
}

std::vector<std::string> split_sql(const std::string& sql) {
    // Strip line comments (-- ... \n), then split on ';\n'.
    std::string clean;
    clean.reserve(sql.size());
    for (std::size_t i = 0; i < sql.size(); ++i) {
        if (sql[i] == '-' && i + 1 < sql.size() && sql[i + 1] == '-') {
            while (i < sql.size() && sql[i] != '\n') ++i;
            if (i < sql.size()) clean.push_back(sql[i]);  // keep the \n
        } else {
            clean.push_back(sql[i]);
        }
    }
    std::vector<std::string> out;
    std::string cur;
    for (std::size_t i = 0; i < clean.size(); ++i) {
        if (clean[i] == ';') {
            // Trim trailing whitespace, then check it's followed by EOL/eof
            std::size_t j = i + 1;
            while (j < clean.size() && (clean[j] == ' ' || clean[j] == '\t')) ++j;
            if (j >= clean.size() || clean[j] == '\n') {
                if (cur.find_first_not_of(" \t\r\n") != std::string::npos) {
                    out.push_back(cur);
                }
                cur.clear();
                i = j;
                continue;
            }
        }
        cur.push_back(clean[i]);
    }
    if (cur.find_first_not_of(" \t\r\n") != std::string::npos) {
        out.push_back(cur);
    }
    return out;
}

}  // namespace forensicator

File: python/forensicator/catalog.py
"""SQLite catalog open/init/upsert helpers."""

from __future__ import annotations

import os
import re
import sqlite3
import sys
import time
from pathlib import Path
from typing import Any, Iterable

from . import SCHEMA_VERSION

_LOCKING_ERROR_PATTERNS = re.compile(
    r"disk I/O error|database is locked|locking protocol", re.IGNORECASE
)


def _find_schema_file() -> Path:
    """Walk up from this file to find schema/forensicator.sql.

    Layout: python/forensicator/catalog.py -> ../../schema/forensicator.sql
    """
    here = Path(__file__).resolve()
    # python/forensicator/catalog.py -> python/forensicator -> python -> repo root
    candidate = here.parent.parent.parent / "schema" / "forensicator.sql"
    if candidate.is_file():
        return candidate
    # Walk further up just in case
    for ancestor in here.parents:
        c = ancestor / "schema" / "forensicator.sql"
        if c.is_file():
            return c
    env = os.environ.get("FORENSICATOR_SCHEMA")
    if env and Path(env).is_file():
        return Path(env)
    raise RuntimeError(
        f"cannot locate schema/forensicator.sql (tried {candidate})"
    )


def schema_sql() -> str:
    return _find_schema_file().read_text(encoding="utf-8")


def _split_sql_statements(sql: str) -> list[str]:
    # Strip line comments
    sql = re.sub(r"--[^\n]*", "", sql)
    parts = re.split(r";\s*(?:\n|$)", sql)
    return [p for p in parts if p.strip()]


def _locking_error_message(path: str, err: str) -> str:
    return (
        f"cannot open catalog {path}: {err}\n"
        "This filesystem appears to have broken POSIX locking (SMB / NFS / exFAT /\n"
        "some FUSE mounts). forensicator refuses to operate on such filesystems\n"
        "because long-running scans corrupt SQLite there (\"database disk image is\n"
        "malformed\"), and the previous nolock fallback caused real data loss.\n"
        "\n"
        "Move the catalog to a local SSD/HDD with a normal filesystem (APFS / ext4\n"
        "/ NTFS / HFS+) and re-run. The catalog records mount_point separately so\n"
        "it can still find files on the original data volume, regardless of where\n"
        "the SQLite file lives.\n"
    )


def open_catalog(
    path: str | os.PathLike,
    *,
    readonly: bool = False,
    dataset: str | None = None,
) -> sqlite3.Connection:
    path = str(path)
    conn = sqlite3.connect(path, isolation_level=None)
    conn.row_factory = sqlite3.Row
    conn.text_factory = str

    if not readonly:
        try:
            conn.execute("PRAGMA journal_mode = WAL")
        except sqlite3.DatabaseError:
            sys.stderr.write(
                f"[forensicator] WAL journal mode unavailable on this filesystem ({path}); using rollback journal\n"
            )
        try:
            conn.execute("PRAGMA synchronous = NORMAL")
        except sqlite3.DatabaseError:
            pass
    try:
        conn.execute("PRAGMA foreign_keys = ON")
    except sqlite3.DatabaseError:
        pass
    try:
        conn.execute("PRAGMA cache_size = -200000")
    except sqlite3.DatabaseError:
        pass

    try:
        has_meta = conn.execute(
            "SELECT name FROM sqlite_master WHERE type='table' AND name='meta'"
        ).fetchone()
        if has_meta:
            sv_raw = catalog_meta(conn, "schema_version")
            sv = int(sv_raw) if sv_raw is not None else 0
            if sv != SCHEMA_VERSION:
                raise RuntimeError(
                    f"catalog {path} has schema version {sv}, expected {SCHEMA_VERSION}"
                )
        if not readonly:
            init_catalog(conn, dataset=dataset)
    except sqlite3.DatabaseError as e:
        try:
            conn.close()
        except Exception:
            pass
        msg = str(e)
        if _LOCKING_ERROR_PATTERNS.search(msg):
            raise RuntimeError(_locking_error_message(path, msg)) from e
        raise RuntimeError(f"cannot open catalog {path}: {msg}") from e
    except Exception as e:
        try:
            conn.close()
        except Exception:
            pass
        msg = str(e)
        if _LOCKING_ERROR_PATTERNS.search(msg):
            raise RuntimeError(_locking_error_message(path, msg)) from e
        raise

    return conn


def init_catalog(conn: sqlite3.Connection, *, dataset: str | None = None) -> None:
    sql = schema_sql()
    for stmt in _split_sql_statements(sql):
        conn.execute(stmt)
    set_catalog_meta(conn, "schema_version", str(SCHEMA_VERSION))
    set_catalog_meta(conn, "created_at", str(int(time.time())))
    if dataset is not None:
        set_catalog_meta(conn, "dataset_name", dataset)


def catalog_meta(conn: sqlite3.Connection, key: str) -> str | None:
    row = conn.execute("SELECT value FROM meta WHERE key = ?", (key,)).fetchone()
    if row is None:
        return None
    return row[0]


def set_catalog_meta(conn: sqlite3.Connection, key: str, value: str) -> None:
    conn.execute(
        "INSERT OR REPLACE INTO meta(key, value) VALUES(?, ?)", (key, value)
    )


def upsert_volume(
    conn: sqlite3.Connection,
    *,
    hostname: str,
    volume: str,
    mount_point: str | None,
    os: str | None,
) -> sqlite3.Row:
    now = int(time.time())
    conn.execute(
        """
        INSERT INTO volumes(hostname, volume, mount_point, os, scanned_at)
        VALUES(?, ?, ?, ?, ?)
        ON CONFLICT(hostname, volume) DO UPDATE SET
          mount_point = excluded.mount_point,
          os          = excluded.os,
          scanned_at  = excluded.scanned_at
        """,
        (hostname, volume, mount_point, os, now),
    )
    return get_volume(conn, hostname, volume)


def get_volume(
    conn: sqlite3.Connection, hostname: str, volume: str
) -> sqlite3.Row | None:
    return conn.execute(
        "SELECT * FROM volumes WHERE hostname = ? AND volume = ?",
        (hostname, volume),
    ).fetchone()


def volumes(conn: sqlite3.Connection) -> list[sqlite3.Row]:
    return list(
        conn.execute("SELECT * FROM volumes ORDER BY hostname, volume")
    )


def begin(conn: sqlite3.Connection) -> None:
    conn.execute("BEGIN")


def commit(conn: sqlite3.Connection) -> None:
    conn.execute("COMMIT")
